federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,244 @@ limitations under the License.
+	if client, ok := s.clusterCache.clientMap[clusterName]; !ok {
+		return nil, "", fmt.Errorf("Cluster cache does not contain entry for cluster %s", clusterName)
+	} else {
+		if client.cluster == nil {
+			return nil, "", fmt.Errorf("Cluster cache entry for cluster %s is nil", clusterName)
+		} else {
+			return client.cluster.Status.Zones, client.cluster.Status.Region, nil
+		}
+	}
 }
 
 // getFederationDNSZoneName returns the name of the managed DNS Zone configured for this federation
-func getFederationDNSZoneName() string {
-	return "mydomain.com" // TODO: quinton: Get this from the federation configuration.
+func (s *ServiceController) getFederationDNSZoneName() (string, error) {
+	return "example.com", nil // TODO: quinton: Get this from the federation configuration.
+	// Note: For unit testing this must match the domain populated in the test/stub dnsprovider.
 }
 
-func ensureDNSRecords(clusterName string, cachedService *cachedService) error {
-	// Quinton: Pseudocode....
+// getZone is a hack around the fact that dnsprovider does not yet support a Get() method, only a List() method.  TODO:  Fix that.
+func getZone(dnsZoneName string, zonesInterface dnsprovider.Zones) (dnsprovider.Zone, error) {
+	zones, err := zonesInterface.List()
+	if err != nil {
+		return nil, err
+	}
+	for _, zone := range zones {
+		if zone.Name() == dnsZoneName {
+			return zone, nil
+		}
+	}
+	return nil, fmt.Errorf("DNS zone %s not found.", dnsZoneName)
+}
+
+/* getRrset is a hack around the fact that dnsprovider.ResourceRecordSets interface does not yet include a Get() method, only a List() method.  TODO:  Fix that.
+   Note that if the named resource record set does not exist, but no error occurred, the returned set, and error, are both nil
+*/
+func getRrset(dnsName string, rrsetsInterface dnsprovider.ResourceRecordSets) (dnsprovider.ResourceRecordSet, error) {
+	var returnVal dnsprovider.ResourceRecordSet
+	rrsets, err := rrsetsInterface.List()
+	if err != nil {
+		return nil, err
+	}
+	for _, rrset := range rrsets {
+		if rrset.Name() == dnsName {
+			returnVal = rrset
+			break
+		}
+	}
+	return returnVal, nil
+}
 
+/* ensureDnsRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resource record sets for dnsName are consistent with endpoints.
+   if endpoints is nil or empty, a CNAME record to uplevelCname is ensured.
+*/
+func (s *ServiceController) ensureDnsRrsets(dnsZoneName, dnsName string, endpoints []string, uplevelCname string) error {
+	dnsZone, err := getZone(dnsZoneName, s.dnsZones)
+	if err != nil {
+		return err
+	}
+	rrsets, supported := dnsZone.ResourceRecordSets()
+	if !supported {
+		return fmt.Errorf("Failed to ensure DNS records for %s. DNS provider does not support the ResourceRecordSets interface.", dnsName)
+	}
+	rrset, err := getRrset(dnsName, rrsets) // TODO: rrsets.Get(dnsName)
+	if err != nil {
+		return err
+	}
+	if rrset == nil {
```

Comment:
> Done in a slightly different way than proposed. getRrset returns (nil, nil) if the record set does not exist (as opposed to if there was an error getting the rrset, in which case err != nil).

> When dnsprovider.ResourceRecordSets is extended to support Get(), we can decide whether to return different types of errors, or adopt the same approach as described above. There are pros and cons to both, but the above approach involves a lot less code now.


What are the pros and cons?

Honestly, defining an error type for NotFound is same amount of code, if not less, as you have now.

```Go
var errNotExist = errors.New("entity does not exist")

...

func getRrset(dnsName string, rrsetsInterface dnsprovider.ResourceRecordSets) (dnsprovider.ResourceRecordSet, error) {
 	rrsets, err := rrsetsInterface.List()
 	if err != nil {
 		return nil, err
 	}
 	for _, rrset := range rrsets {
 		if rrset.Name() == dnsName {
 			return rrset, nil
		}
	}
	return nil, errNotExist
}

func (s *ServiceController) ensureDnsRrsets(dnsZoneName, dnsName string, endpoints []string, uplevelCname string) error {
	...
	rrset, err := getRrset(dnsName, rrsets) // TODO: rrsets.Get(dnsName)
	if err != nil && err != errNotExist {
		return err
	} else if err == errNotExist {
		// Do what needs to be done when rrset doesn't exist
	} else {
		// Do what needs to be done when rrset exists.
	}
	...
}
...

The problem with the current approach is that, there is no explicit contract between the caller and the callee. This proposed approach establishes that contract and makes it explicit. As a result the code becomes lot more clear to the reader. This logic is effectively a 3-state switch and should be represented that way for clarity.